include/swift-c/DependencyScan/DependencyScan.h

```diff
@@ -441,6 +441,52 @@ typedef struct swiftscan_cas_options_s *swiftscan_cas_options_t;
 /// ActionCache.
 typedef struct swiftscan_cas_s *swiftscan_cas_t;
 
+/// Opaque container for a cached compilation.
+typedef struct swiftscan_cached_compilation_s *swiftscan_cached_compilation_t;
+
+/// Opaque container for a cached compilation output.
+typedef struct swiftscan_cached_output_s *swiftscan_cached_output_t;
+
+/// Opaque type for a cache replay instance.
+typedef struct swiftscan_cache_replay_instance_s
+    *swiftscan_cache_replay_instance_t;
+
+/// Opaque container for a cached compilation replay result.
+typedef struct swiftscan_cache_replay_result_s *swiftscan_cache_replay_result_t;
+
+/// Opaque type for a cancellation token for async cache operations.
+typedef struct swiftscan_cache_cancellation_token_s
+    *swiftscan_cache_cancellation_token_t;
+
+/// Enum types for output types for cache key computation.
+/// List should contain all the output file types, including supplemententary
+/// outputs, except diagnostics outputs, which are covered by cached diagnostic
+/// entry.
+typedef enum {
+  SWIFTSCAN_OUTPUT_TYPE_OBJECT = 0,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFTMODULE = 1,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFTINTERFACE = 2,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFTPRIVATEINTERFACE = 3,
+  SWIFTSCAN_OUTPUT_TYPE_CLANG_MODULE = 4,
+  SWIFTSCAN_OUTPUT_TYPE_CLANG_PCH = 5,
+  SWIFTSCAN_OUTPUT_TYPE_CLANG_HEADER = 6,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFT_SOURCE_INFO = 7,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFT_MODULE_DOC = 8,
+  SWIFTSCAN_OUTPUT_TYPE_DEPENDENCIES = 9,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFT_DEPS = 10,
+  SWIFTSCAN_OUTPUT_TYPE_MODULE_TRACE = 11,
+  SWIFTSCAN_OUTPUT_TYPE_TBD = 12,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFT_MODULE_SUMMARY = 13,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFT_ABI_DESCRIPTOR = 14,
+  SWIFTSCAN_OUTPUT_TYPE_SWIFT_API_DESCRIPTOR = 15,
+  SWIFTSCAN_OUTPUT_TYPE_CONST_VALUE = 16,
+  SWIFTSCAN_OUTPUT_TYPE_MODULE_SEMANTIC_INFO = 17,
+  SWIFTSCAN_OUTPUT_TYPE_YAML_OPT_RECORD = 18,
+  SWIFTSCAN_OUTPUT_TYPE_BITSTREAM_OPT_RECORD = 19,
+  SWIFTSCAN_OUTPUT_TYPE_CACHED_DIAGNOSTICS = 20,
+  SWIFTSCAN_OUTPUT_TYPE_LAST = SWIFTSCAN_OUTPUT_TYPE_CACHED_DIAGNOSTICS
+} swiftscan_output_kind_t;
```

Comment:
This allows swift-driver to perform the reverse mapping its `FileType`: https://github.com/apple/swift-driver/blob/eefc6581e50adf1ba6d492f4e023a824ab06dee8/Sources/SwiftDriver/SwiftScan/SwiftScanCAS.swift#L107

This allows swift-driver to associate the kind with the output `FileType` it knows about for the job it is trying to create. swift-driver also knows how to print `FileType` so I think this is better than just a name.

Reply:
I'm still unclear why the kind enums are useful, doesn't the swift-driver already know what the outputs are from the invocation arguments? Why does it need to get the kinds from the cached compilation?

Reply:
For the enums, it allows it to tell which cached output is corresponding output from its computed job. It is probably not useful for current swift driver but if we need to convey the output kind, I prefer it to be a enum than a string. If we ever need the enum, it will render the string version useless. The downside is that we need a stable ordering for the enum value, which I don't see it being too big a problem.

Reply:
You are creating an API that ties the client implementation with the compiler implementation, so it has a downside without providing any benefit currently.

If we have a function that just provides the name, the client can log it, and if later the compiler implementation adds a new output the client doesn't need to change anything in its code to be able to log the name of the new output.
But using an enum means we have to coordinate and sync-up both the client and the compiler in order to be able to log the name of a new output.

We can always add the kind back in later, when there is a benefit so the tradeoff of tying the implementations is worth it.

Reply:
The outputs needed from the compilation is first computed by driver, which send to frontend as command-line argument. It is currently tied together already and frontend should not try to produce output that are not expected by driver.